kore/src/Kore/Step/Simplification/Implies.hs

```diff
@@ -88,9 +88,11 @@ simplifyEvaluated sideCondition first second
   | OrPattern.isFalse first  = return OrPattern.top
   | OrPattern.isTrue second  = return OrPattern.top
   | OrPattern.isFalse second = Not.simplifyEvaluated sideCondition first
-  | otherwise = do
-    results <- traverse (simplifyEvaluateHalfImplies sideCondition first) second
-    return (MultiOr.flatten results)
+  | otherwise =
+      OrPattern.flatten
+      <$> OrPattern.traverse
+          (simplifyEvaluateHalfImplies sideCondition first)
+          second
```

Comment:
This would be a good place to use `observeAllT` instead of `flatten`:

```suggestion
  | otherwise =
      OrPattern.observeAllT $ do
          second' <- Logic.scatter second
          simplified <- simplifyEvaluateHalfImplies sideCondition first second'
          Logic.scatter simplified
```

If this is the only call site of `simplifyEvaluateHalfImplies` (I think it is) then you can also change the signature of that function to return a `LogicT _ _` and simplify this function further. I'm not sure that `OrPattern.flatten` is actually necessary at all.